unittest.TestCase cleanup functions not invoked on test failure

stdlib unittest style cleanup functions registered with `unittest.TestCase.addCleanup` are not invoked when a test fails.  It appears this issue was introduced in pytest version 5.4.0, examples below for version 5.4.1.

### System Info

- Ubuntu 18.04.3 LTS
- Python 3.6.8
- pytest 5.4.1


### Example test and pytest output

```python
import unittest

def cleanup():
    raise Exception('cleanup')

class Test(unittest.TestCase):
    def setUp(self):
        print('setup')
        self.addCleanup(cleanup)

    def tearDown(self):
        print('teardown')

    def test_no_cleanup(self):
        assert False

    def test_cleanup(self):
        assert True

```

```
(venv-3.6.8) cecil@python36-vm:~$ pytest ceciltest.py
========================================================================================= test session starts ==========================================================================================
platform linux -- Python 3.6.8, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/cecil
plugins: cov-2.8.1
collected 2 items

ceciltest.py FF                                                                                                                                                                                  [100%]

=============================================================================================== FAILURES ===============================================================================================
__________________________________________________________________________________________ Test.test_cleanup ___________________________________________________________________________________________

    def cleanup():
>       raise Exception('cleanup!')
E       Exception: cleanup!

ceciltest.py:4: Exception
----------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------
setup
teardown
_________________________________________________________________________________________ Test.test_no_cleanup _________________________________________________________________________________________

self = <ceciltest.Test testMethod=test_no_cleanup>

    def test_no_cleanup(self):
>       assert False
E       assert False

ceciltest.py:16: AssertionError
----------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------
setup
--------------------------------------------------------------------------------------- Captured stdout teardown ---------------------------------------------------------------------------------------
teardown
======================================================================================= short test summary info ========================================================================================
FAILED ceciltest.py::Test::test_cleanup - Exception: cleanup!
FAILED ceciltest.py::Test::test_no_cleanup - assert False
========================================================================================== 2 failed in 0.12s ===========================================================================================

```

### Trying pytest 5.3.5 (works as expected)

```
(venv-3.6.8) cecil@python36-vm:~$ pytest ceciltest.py
========================================================================================= test session starts ==========================================================================================
platform linux -- Python 3.6.8, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /home/cecil
plugins: cov-2.8.1
collected 2 items

ceciltest.py FFE                                                                                                                                                                                 [100%]

================================================================================================ ERRORS ================================================================================================
______________________________________________________________________________ ERROR at teardown of Test.test_no_cleanup _______________________________________________________________________________

    def cleanup():
>       raise Exception('cleanup!')
E       Exception: cleanup!

ceciltest.py:4: Exception
----------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------
setup
teardown
=============================================================================================== FAILURES ===============================================================================================
__________________________________________________________________________________________ Test.test_cleanup ___________________________________________________________________________________________

    def cleanup():
>       raise Exception('cleanup!')
E       Exception: cleanup!

ceciltest.py:4: Exception
----------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------
setup
teardown
_________________________________________________________________________________________ Test.test_no_cleanup _________________________________________________________________________________________

self = <ceciltest.Test testMethod=test_no_cleanup>

    def test_no_cleanup(self):
>       assert False
E       AssertionError: assert False

ceciltest.py:16: AssertionError
----------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------
setup
teardown
====================================================================================== 2 failed, 1 error in 0.12s ======================================================================================
```

### pip list
```
(venv-3.6.8) cecil@python36-vm:~$ pip list
Package            Version
------------------ ----------
astroid            2.3.3
attrs              19.3.0
boto               2.49.0
certifi            2019.11.28
chardet            3.0.4
coverage           5.0.4
httpretty          0.8.10
idna               2.9
importlib-metadata 1.5.0
isort              4.3.21
Jinja2             2.11.1
lazy-object-proxy  1.4.3
MarkupSafe         1.1.1
mccabe             0.6.1
more-itertools     8.2.0
moto               0.4.31
packaging          20.3
parameterized      0.7.0
pep8               1.6.2
pip                18.1
pluggy             0.13.1
py                 1.8.1
pylint             2.4.4
pyparsing          2.4.6
pytest             5.4.1
pytest-cov         2.8.1
python-dateutil    2.8.1
pytz               2019.3
requests           2.23.0
setuptools         40.6.2
six                1.14.0
typed-ast          1.4.1
urllib3            1.25.8
wcwidth            0.1.8
Werkzeug           1.0.0
wrapt              1.11.2
xmltodict          0.12.0
zipp               3.1.0

```

--- a/changelog/6947.bugfix.rst
+++ b/changelog/6947.bugfix.rst
@@ -1,0 +1,1 @@
+Fix regression where functions registered with ``TestCase.addCleanup`` were not being called on test failures.

--- a/src/_pytest/debugging.py
+++ b/src/_pytest/debugging.py
@@ -272,11 +272,15 @@ class PdbInvoke:
 class PdbTrace:
     @hookimpl(hookwrapper=True)
     def pytest_pyfunc_call(self, pyfuncitem):
-        _test_pytest_function(pyfuncitem)
+        wrap_pytest_function_for_tracing(pyfuncitem)
         yield
 
 
-def _test_pytest_function(pyfuncitem):
+def wrap_pytest_function_for_tracing(pyfuncitem):
+    """Changes the python function object of the given Function item by a wrapper which actually
+    enters pdb before calling the python function itself, effectively leaving the user
+    in the pdb prompt in the first statement of the function.
+    """
     _pdb = pytestPDB._init_pdb("runcall")
     testfunction = pyfuncitem.obj
 
@@ -289,6 +293,13 @@ def _test_pytest_function(pyfuncitem):
         _pdb.runcall(func)
 
     pyfuncitem.obj = wrapper
+
+
+def maybe_wrap_pytest_function_for_tracing(pyfuncitem):
+    """Wrap the given pytestfunct item for tracing support if --trace was given in
+    the command line"""
+    if pyfuncitem.config.getvalue("trace"):
+        wrap_pytest_function_for_tracing(pyfuncitem)
 
 
 def _enter_pdb(node, excinfo, rep):

--- a/src/_pytest/unittest.py
+++ b/src/_pytest/unittest.py
@@ -1,5 +1,4 @@
 """ discovery and running of std-library "unittest" style tests. """
-import functools
 import sys
 import traceback
 
@@ -114,15 +113,17 @@ class TestCaseFunction(Function):
     _testcase = None
 
     def setup(self):
-        self._needs_explicit_tearDown = False
+        # a bound method to be called during teardown() if set (see 'runtest()')
+        self._explicit_tearDown = None
         self._testcase = self.parent.obj(self.name)
         self._obj = getattr(self._testcase, self.name)
         if hasattr(self, "_request"):
             self._request._fillfixtures()
 
     def teardown(self):
-        if self._needs_explicit_tearDown:
-            self._testcase.tearDown()
+        if self._explicit_tearDown is not None:
+            self._explicit_tearDown()
+            self._explicit_tearDown = None
         self._testcase = None
         self._obj = None
 
@@ -205,40 +206,31 @@ class TestCaseFunction(Function):
         return bool(expecting_failure_class or expecting_failure_method)
 
     def runtest(self):
-        # TODO: move testcase reporter into separate class, this shouldnt be on item
-        import unittest
+        from _pytest.debugging import maybe_wrap_pytest_function_for_tracing
 
-        testMethod = getattr(self._testcase, self._testcase._testMethodName)
-
-        class _GetOutOf_testPartExecutor(KeyboardInterrupt):
-            """Helper exception to get out of unittests's testPartExecutor (see TestCase.run)."""
-
-        @functools.wraps(testMethod)
-        def wrapped_testMethod(*args, **kwargs):
-            """Wrap the original method to call into pytest's machinery, so other pytest
-            features can have a chance to kick in (notably --pdb)"""
-            try:
-                self.ihook.pytest_pyfunc_call(pyfuncitem=self)
-            except unittest.SkipTest:
-                raise
-            except Exception as exc:
-                expecting_failure = self._expecting_failure(testMethod)
-                if expecting_failure:
-                    raise
-                self._needs_explicit_tearDown = True
-                raise _GetOutOf_testPartExecutor(exc)
+        maybe_wrap_pytest_function_for_tracing(self)
 
         # let the unittest framework handle async functions
         if is_async_function(self.obj):
             self._testcase(self)
         else:
-            setattr(self._testcase, self._testcase._testMethodName, wrapped_testMethod)
+            # when --pdb is given, we want to postpone calling tearDown() otherwise
+            # when entering the pdb prompt, tearDown() would have probably cleaned up
+            # instance variables, which makes it difficult to debug
+            # arguably we could always postpone tearDown(), but this changes the moment where the
+            # TestCase instance interacts with the results object, so better to only do it
+            # when absolutely needed
+            if self.config.getoption("usepdb"):
+                self._explicit_tearDown = self._testcase.tearDown
+                setattr(self._testcase, "tearDown", lambda *args: None)
+
+            # we need to update the actual bound method with self.obj, because
+            # wrap_pytest_function_for_tracing replaces self.obj by a wrapper
+            setattr(self._testcase, self.name, self.obj)
             try:
                 self._testcase(result=self)
-            except _GetOutOf_testPartExecutor as exc:
-                raise exc.args[0] from exc.args[0]
             finally:
-                delattr(self._testcase, self._testcase._testMethodName)
+                delattr(self._testcase, self.name)
 
     def _prunetraceback(self, excinfo):
         Function._prunetraceback(self, excinfo)

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -876,6 +876,37 @@ def test_no_teardown_if_setupclass_failed(testdir):
     reprec.assertoutcome(passed=1, failed=1)
 
 
+def test_cleanup_functions(testdir):
+    """Ensure functions added with addCleanup are always called after each test ends (#6947)"""
+    testdir.makepyfile(
+        """
+        import unittest
+
+        cleanups = []
+
+        class Test(unittest.TestCase):
+
+            def test_func_1(self):
+                self.addCleanup(cleanups.append, "test_func_1")
+
+            def test_func_2(self):
+                self.addCleanup(cleanups.append, "test_func_2")
+                assert 0
+
+            def test_func_3_check_cleanups(self):
+                assert cleanups == ["test_func_1", "test_func_2"]
+    """
+    )
+    result = testdir.runpytest("-v")
+    result.stdout.fnmatch_lines(
+        [
+            "*::test_func_1 PASSED *",
+            "*::test_func_2 FAILED *",
+            "*::test_func_3_check_cleanups PASSED *",
+        ]
+    )
+
+
 def test_issue333_result_clearing(testdir):
     testdir.makeconftest(
         """
@@ -1129,6 +1160,41 @@ def test_trace(testdir, monkeypatch):
     result = testdir.runpytest("--trace", str(p1))
     assert len(calls) == 2
     assert result.ret == 0
+
+
+def test_pdb_teardown_called(testdir, monkeypatch):
+    """Ensure tearDown() is always called when --pdb is given in the command-line.
+
+    We delay the normal tearDown() calls when --pdb is given, so this ensures we are calling
+    tearDown() eventually to avoid memory leaks when using --pdb.
+    """
+    teardowns = []
+    monkeypatch.setattr(
+        pytest, "test_pdb_teardown_called_teardowns", teardowns, raising=False
+    )
+
+    testdir.makepyfile(
+        """
+        import unittest
+        import pytest
+
+        class MyTestCase(unittest.TestCase):
+
+            def tearDown(self):
+                pytest.test_pdb_teardown_called_teardowns.append(self.id())
+
+            def test_1(self):
+                pass
+            def test_2(self):
+                pass
+    """
+    )
+    result = testdir.runpytest_inprocess("--pdb")
+    result.stdout.fnmatch_lines("* 2 passed in *")
+    assert teardowns == [
+        "test_pdb_teardown_called.MyTestCase.test_1",
+        "test_pdb_teardown_called.MyTestCase.test_2",
+    ]
 
 
 def test_async_support(testdir):

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -537,28 +537,24 @@ class TestTrialUnittest:
         )
         result.stdout.fnmatch_lines(
             [
-                "test_trial_error.py::TC::test_four SKIPPED",
+                "test_trial_error.py::TC::test_four FAILED",
                 "test_trial_error.py::TC::test_four ERROR",
                 "test_trial_error.py::TC::test_one FAILED",
                 "test_trial_error.py::TC::test_three FAILED",
-                "test_trial_error.py::TC::test_two SKIPPED",
-                "test_trial_error.py::TC::test_two ERROR",
+                "test_trial_error.py::TC::test_two FAILED",
                 "*ERRORS*",
                 "*_ ERROR at teardown of TC.test_four _*",
-                "NOTE: Incompatible Exception Representation, displaying natively:",
-                "*DelayedCalls*",
-                "*_ ERROR at teardown of TC.test_two _*",
-                "NOTE: Incompatible Exception Representation, displaying natively:",
                 "*DelayedCalls*",
                 "*= FAILURES =*",
-                # "*_ TC.test_four _*",
-                # "*NameError*crash*",
+                "*_ TC.test_four _*",
+                "*NameError*crash*",
                 "*_ TC.test_one _*",
                 "*NameError*crash*",
                 "*_ TC.test_three _*",
-                "NOTE: Incompatible Exception Representation, displaying natively:",
                 "*DelayedCalls*",
-                "*= 2 failed, 2 skipped, 2 errors in *",
+                "*_ TC.test_two _*",
+                "*NameError*crash*",
+                "*= 4 failed, 1 error in *",
             ]
         )
 

